Nested ldap groups

CarlSchwan opened the pull request:
This PR adds support for nested groups and also increases the amount of caching. This can be tested with https://github.com/juliushaertl/nextcloud-docker-dev/pull/51

Please look at the different commit messages for more explanation of every change

CarlSchwan committed:
Unify a bit the types of the fetcher

Now it will only accept a string as parameter instead of either a string
(DN) or a array (complete record).

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Add more type hinting

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Fix merging list with null

This fixes some cases observed with the debugger where we end up merging
a non empty list with null. The result is then null and the looping over
the items would then end.

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

blizzz committed:
Make it possible to return nested records whem walking over groups

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
Refactor code to split common loop

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
Cache intermediates

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>
Co-authored-by: Carl Schwan <carl@carlschwan.eu>

come-nc reviewed (commented):
I am still against this solution of caching differently intermediate result instead of fixing them.
Fixing them might actually improve performances as they would be cached accross requests like other groups.

come-nc committed:
Refactor _groupMembers to correctly use cache on intermediate results

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> I am still against this solution of caching differently intermediate result instead of fixing them. Fixing them might actually improve performances as they would be cached accross requests like other groups.

@CarlSchwan https://github.com/nextcloud/server/commit/c21c7f2c2520fde5daf4755db1bc9b53e3c0e4d5 What do you think of this for instance?
I have 2 failing tests but they are failing before my changes as well.

CarlSchwan commented:
> > I am still against this solution of caching differently intermediate result instead of fixing them. Fixing them might actually improve performances as they would be cached accross requests like other groups.
> 
> @CarlSchwan [c21c7f2](https://github.com/nextcloud/server/commit/c21c7f2c2520fde5daf4755db1bc9b53e3c0e4d5) What do you think of this for instance? I have 2 failing tests but they are failing before my changes as well.

The patch looks good and also easier to read and understand

come-nc committed:
Small optimisation of _groupMembers

This will not change the result as users are check to be existing
 afterwards but avoids this check when we know it’s a group.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add tests for nested groups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Refactor group membership listing for nested groups

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add testing of nested group membership

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Fix types in docblocks

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Removed unused use declaration

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
Add missing copyright author in Group_LDAP

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Rebased on master and solved conflicts

CarlSchwan commented:
Rebased this PR and found some issues :(

With two groups that are circular: A (with 3 members) and B (with 2 members)
 
1. when loading the list of the member of A, we see 5 members, this is correct
2. when loading the list of the member of B, we only see 2 members, this is not correct
3. Clearing the redis cache
4. when loading the list of the member of B, we only see 5 members, this is correct
5. when loading the list of the member of A, we see 3 members, this is not correct

This was fixed with https://github.com/nextcloud/server/commit/682a6acd7da6ef8251c3c4430f337152a2907f77 and https://github.com/nextcloud/server/commit/4e165b1e8583c48fe8b7b8c575277d4afbad9009 in the stable20 fork

CarlSchwan committed:
Only cache base inGroup search

And not intermediate search for nested groups, this is causing issues
othewise with nested groups

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Add back runtime cache for intermediate ldap read results

This is a small optimization that save a few LDAP queries

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan commented:
Found the issue! This 

```
$shouldCacheResult = count($seen) === 0;
```

got removed and when we ended up caching the intermediate steps, something that works fine for nested groups but not with circular groups. Now locally this patch seems to work.

I added it back and also added back the runtime cache optimization

CarlSchwan committed:
Fix user_ldap tests

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan committed:
Revert Carl changes on apps/user_ldap/lib/Group_LDAP.php

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

come-nc commented on apps/user_ldap/lib/Group_LDAP.php:
```diff
@@ -808,7 +808,7 @@ private function getGroupsByMember(string $dn, array &$seen = []): array {
 			return [];
 		}
-		$cacheKey = $dn . implode('-', array_keys($seen));
+		$cacheKey = $dn . implode('-', $seen);
```
This makes no sense, values in $seen are booleans, imploding the values cannot produce a meaningful thing.

come-nc commented:
@CarlSchwan One of the points of the refactor as I remember it was to be able to cache all steps, so https://github.com/nextcloud/server/pull/30223/commits/07dec64ea991a4284743b0e20c31cd46dcfd497e looks wrong to me.

Can you provide a test case that fails without this commit?

come-nc committed:
Fix LDAP recursive nested group support

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Update on this:

There was indeed a caching bug left in my code in the special case of recursive nested groups. We came to the conclusion that we have to not cache intermediate results when recursion is detected. This allows to cache all results in normal situations and only avoid caching for recursive nested groups.

This kind of bugs are not seen by tests because the tests use a mockup that does not do caching. We need to fix this and add tests that instanciate the caching so that these bugs are visible.

I have a local patch that fixes the bug but I need to coordinate with @CarlSchwan to see which part of his last commits needs reverting, and we need to test getting groups of a user, because my patch only affect getting members of a group.
Once we clear that out and have proper tests for caching, I think we can merge. We should also check what the profiler says about before VS after merging this.

come-nc committed:
Bring back small fixes by Carl

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
/rebase

CarlSchwan committed:
Check if cache is present with isset

Otherwise we get false for empty array

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

CarlSchwan commented:
Performance before:
- DB request: 120
- LDAP request: 169

Performance after:
- DB request: 102
- LDAP request: 103

In both case with a cold cache

CarlSchwan committed:
Fix psalm

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

come-nc reviewed (approved):
:+1: Tested and validated with Carl
